apps/service_providers/models.py

```diff
@@ -120,14 +128,71 @@ def get_llm_service(self):
         return self.type_enum.get_llm_service(config)
 
 
+class LlmProviderModelManager(models.Manager):
+    def for_team(self, team):
+        return super().get_queryset().filter(models.Q(team=team) | models.Q(team__isnull=True))
+
+
+class LlmProviderModel(BaseTeamModel):
+    team = models.ForeignKey(
+        Team,
+        verbose_name=gettext("Team"),
+        on_delete=models.CASCADE,
+        null=True,
+        blank=True,
+    )  # Optional FK relationship to team. If this is
+    # null, then it is a "global" model that is managed in Django admin. Team,
+
+    type = models.CharField(max_length=255, choices=LlmProviderTypes.choices)
+
+    name = models.CharField(
+        max_length=128, help_text="The name of the model. e.g. 'gpt-4o-mini' or 'claude-3-5-sonnet-20240620'"
+    )
+    max_token_limit = models.PositiveIntegerField(
+        default=8192,
+        help_text="When the message history for a session exceeds this limit (in tokens), it will be compressed. "
+        "If 0, compression will be disabled which may result in errors or high LLM costs.",
+    )
+
+    objects = LlmProviderModelManager()
+
+    class Meta:
+        constraints = [
+            models.UniqueConstraint(
+                fields=("team", "name", "type", "max_token_limit"), name="unique_team_name_type_max_token_limit"
+            ),
+        ]
+
+    def __str__(self):
+        label = f"{LlmProviderTypes[self.type].label}: {self.name}"
+        if self.is_custom():
+            label = f"{label} (custom for {self.team.name})"
+        return label
+
+    def is_custom(self):
+        return self.team is not None
+
+    def delete(self, *args, **kwargs):
```

Comment:
Just noting that we said we'd do a similar check when editing to prevent changing the name of a model in use.

Also, we'd need to check more than just experiments right? For Django model references we could do this:
```python
for field in self._meta.get_fields():
    if field.one_to_many and field.auto_created:
        related_objects = getattr(instance, field.get_accessor_name(), None)
        if related_objects and related_objects.exists():
            raise DjangoValidationError(f"The model '{instance.name}' is still referenced by other objects.")
```

We'd have to manually check pipelines unless setup up FK fields.

Reply:
https://github.com/dimagi/open-chat-studio/pull/810/commits/b82b3ecd55030ce39fe590b7b8f6c4f2e79faa6d